File: core/python/tests/test_spec.py
from types import SimpleNamespace

from pubready.spec import default_appearance, spec_from_args


def test_spec_from_args_builds_python_spec_with_defaults():
    a = SimpleNamespace(recipe="two_group_compare", data="/p/tumor_volume.csv",
                        x="group", y="volume", test="auto")
    spec = spec_from_args(a)
    assert spec["engine"] == "python"
    assert spec["recipe"] == "two_group_compare"
    assert spec["data"]["file"] == "tumor_volume.csv"
    assert spec["data"]["x"] == "group"
    assert spec["test"]["method"] == "auto"
    assert spec["test"]["paired"] is False
    assert spec["appearance"]["geom"] == "box"


def test_spec_from_args_splits_palette_and_covariates():
    a = SimpleNamespace(recipe="cox_forest", data="trial.csv", x=None, y=None,
                        palette="#3B6DB3, #C1432B", covariates="age, sex, stage")
    spec = spec_from_args(a)
    assert spec["appearance"]["palette"] == ["#3B6DB3", "#C1432B"]
    assert spec["data"]["covariates"] == ["age", "sex", "stage"]


def test_default_appearance_shape():
    ap = default_appearance()
    assert ap["bracket"] == {"show": True, "label": "p.signif"}
    assert ap["y_limits"] == [None, None]


def test_spec_from_args_carries_theme():
    a = SimpleNamespace(recipe="two_group_compare", data="d.csv",
                        x="g", y="v", theme="prism")
    assert spec_from_args(a)["appearance"]["theme"] == "prism"
    b = SimpleNamespace(recipe="two_group_compare", data="d.csv", x="g", y="v")
    assert spec_from_args(b)["appearance"]["theme"] == "pubready_house"


def test_spec_from_args_carries_new_data_and_cutoff_fields():
    a = SimpleNamespace(recipe="pca", data="d.csv", x=None, y=None, id="subj",
                        group="grp", label="gene", fc_cutoff=2.0, p_cutoff=0.01, top_n=20)
    s = spec_from_args(a)
    assert s["data"]["id"] == "subj"
    assert s["data"]["group"] == "grp"
    assert s["data"]["label"] == "gene"
    assert s["appearance"]["fc_cutoff"] == 2.0
    assert s["appearance"]["p_cutoff"] == 0.01
    assert s["appearance"]["top_n"] == 20

File: core/r/tests/testthat/test-spec.R
test_that("spec_from_opt builds the expected spec with house defaults", {
  opt <- list(recipe = "two_group_compare", data = "/path/to/tumor_volume.csv",
              x = "group", y = "volume", test = "auto", paired = FALSE,
              p_adjust = "none")
  spec <- spec_from_opt(opt)

  expect_equal(spec$engine, "r")
  expect_equal(spec$recipe, "two_group_compare")
  expect_equal(spec$data$file, "tumor_volume.csv")
  expect_equal(spec$data$x, "group")
  expect_equal(spec$data$y, "volume")
  expect_equal(spec$test$method, "auto")
  expect_false(spec$test$paired)
  expect_equal(spec$appearance$geom, "box")
  expect_true(spec$appearance$show_points)
})

test_that("spec_from_opt splits palette and covariates on commas", {
  opt <- list(recipe = "cox_forest", data = "trial.csv", palette = "#3B6DB3, #C1432B",
              covariates = "age, sex, stage", test = "auto")
  spec <- spec_from_opt(opt)
  expect_equal(spec$appearance$palette, c("#3B6DB3", "#C1432B"))
  expect_equal(spec$data$covariates, c("age", "sex", "stage"))
})

test_that("spec_from_opt carries the theme and defaults to the house style", {
  opt <- list(recipe = "two_group_compare", data = "d.csv", x = "g", y = "v",
              theme = "prism")
  expect_equal(spec_from_opt(opt)$appearance$theme, "prism")
  opt2 <- list(recipe = "two_group_compare", data = "d.csv", x = "g", y = "v")
  expect_equal(spec_from_opt(opt2)$appearance$theme, "pubplot_house")
})

test_that("fix_axis_keys restores YAML-coerced boolean axis keys", {
  d <- list("TRUE" = "volume", x = "group")
  fixed <- fix_axis_keys(d)
  expect_true("y" %in% names(fixed))
  expect_false("TRUE" %in% names(fixed))
  expect_equal(fixed$y, "volume")
})
